src/include/platform/OTARequestorInterface.h

```diff
@@ -30,6 +30,115 @@
 
 namespace chip {
 
+/**
+ * A class to represent a list of the provider locations
+ */
+class ProviderLocationList
+{
+public:
+    /**
+     * A class to iterate over an instance of ProviderLocationList
+     */
+    class Iterator
+    {
+    public:
+        /**
+         * Initialize iterator values, must be called before calling Next()/GetValue()
+         */
+        Iterator(const Optional<app::Clusters::OtaSoftwareUpdateRequestor::Structs::ProviderLocation::Type> * const list,
+                 size_t size)
+        {
+            mList         = list;
+            mListSize     = size;
+            mCurrentIndex = 0;
+            mNextIndex    = 0;
+        }
+
+        /**
+         * Search for the next provider location found in the list
+         */
+        bool Next()
+        {
+            for (size_t i = mNextIndex; i < mListSize; i++)
+            {
+                if (mList[i].HasValue())
+                {
+                    mCurrentIndex = i;
+                    mNextIndex    = mCurrentIndex + 1;
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        /**
+         * Retrieves a reference to the provider location found on a previous call to Next()
+         */
+        const app::Clusters::OtaSoftwareUpdateRequestor::Structs::ProviderLocation::Type & GetValue() const
+        {
+            return mList[mCurrentIndex].Value();
+        }
+
+    private:
+        const Optional<app::Clusters::OtaSoftwareUpdateRequestor::Structs::ProviderLocation::Type> * mList;
+        size_t mListSize;
+        size_t mCurrentIndex;
+        size_t mNextIndex;
+    };
+
+    /**
+     * Retrieve an iterator to the list
+     */
+    Iterator Begin() const { return Iterator(mList, CHIP_CONFIG_MAX_FABRICS); }
+
+    /**
+     * Add a provider location to the list if there is space available
+     */
+    CHIP_ERROR Add(const app::Clusters::OtaSoftwareUpdateRequestor::Structs::ProviderLocation::Type & providerLocation)
+    {
+        for (size_t i = 0; i < mMaxSize; i++)
+        {
+            if (!mList[i].HasValue())
+            {
+                mList[i].SetValue(providerLocation);
+                mListSize++;
+                return CHIP_NO_ERROR;
+            }
+        }
+
+        return CHIP_ERROR_NO_MEMORY;
+    }
+
+    /**
+     * Delete a provider location from the list if found
+     */
+    CHIP_ERROR Delete(const app::Clusters::OtaSoftwareUpdateRequestor::Structs::ProviderLocation::Type & providerLocation)
+    {
+        for (size_t i = 0; i < mMaxSize; i++)
+        {
+            if (mList[i].HasValue())
+            {
+                app::Clusters::OtaSoftwareUpdateRequestor::Structs::ProviderLocation::Type pl = mList[i].Value();
```

Comment:
Does this need to make a copy, or should it be declared as a reference?

Reply:
This does not need to be a copy, I'll make it a reference